### DISCUSSION TITLE
The memory block 'Elsa.Expressions.Models.MemoryBlockReference' does not exist

### DISCUSSION THREAD

**Discussion by mremadevi** [elsa-workflows/elsa-core#5287] [Q&A]
Getting error The memory block 'Elsa.Expressions.Models.MemoryBlockReference' does not exist for below workflow. Please advise.

{
	"id": "19063866e115b482",
	"definitionId": "5a9c5d84332151a5",
	"name": "Workflow 2",
	"createdAt": "2024-04-26T18:03:01.402299+00:00",
	"version": 1,
	"toolVersion": "3.1.0.0",
	"variables": [],
	"inputs": [],
	"outputs": [],
	"outcomes": [],
	"customProperties": {},
	"isReadonly": false,
	"isSystem": false,
	"isLatest": true,
	"isPublished": false,
	"options": {
		"autoUpdateConsumingWorkflows": false
	},
	"root": {
		"type": "Elsa.Flowchart",
		"version": 1,
		"id": "626874ef88bcf26c",
		"nodeId": "Workflow1:626874ef88bcf26c",
		"metadata": {},
		"customProperties": {
			"source": "FlowchartJsonConverter.cs:45",
			"notFoundConnections": [],
			"canStartWorkflow": false,
			"runAsynchronously": false
		},
		"activities": [
			{
				"result": null,
				"id": "5212ab931be19aee",
				"nodeId": "Workflow1:626874ef88bcf26c:5212ab931be19aee",
				"name": "ActivityWithOutput1",
				"type": "Wave.Service.HostedServiceHandlers.ExternalEvents.WaveDrafted.ElsaWorkflow.ActivityWithOutput",
				"version": 1,
				"customProperties": {
					"canStartWorkflow": false,
					"runAsynchronously": false
				},
				"metadata": {
					"designer": {
						"position": {
							"x": 153.5,
							"y": 304
						},
						"size": {
							"width": 174.46875,
							"height": 50
						}
					}
				}
			},
			{
				"condition": {
					"typeName": "Boolean",
					"expression": {
						"type": "CSharp",
						"value": "(string)Output.LastResult == \"test\""
					}
				},
				"then": {
					"type": "Elsa.Flowchart",
					"version": 1,
					"id": "c90bc46199ca62bb",
					"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:c90bc46199ca62bb",
					"metadata": {},
					"customProperties": {
						"source": "FlowchartJsonConverter.cs:45",
						"notFoundConnections": [],
						"canStartWorkflow": false,
						"runAsynchronously": false
					},
					"activities": [
						{
							"result": null,
							"id": "3d5b9ca8dde0b9d4",
							"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:c90bc46199ca62bb:3d5b9ca8dde0b9d4",
							"name": "CommitWaveActivity1",
							"type": "Wave.Service.HostedServiceHandlers.ExternalEvents.WaveDrafted.ElsaWorkflow.CommitWaveActivity",
							"version": 1,
							"customProperties": {
								"canStartWorkflow": false,
								"runAsynchronously": false
							},
							"metadata": {
								"designer": {
									"position": {
										"x": -474.5,
										"y": -221
									},
									"size": {
										"width": 189.625,
										"height": 50
									}
								}
							}
						}
					],
					"connections": []
				},
				"else": {
					"type": "Elsa.Flowchart",
					"version": 1,
					"id": "a2307a0b4b92bb79",
					"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:a2307a0b4b92bb79",
					"metadata": {},
					"customProperties": {
						"source": "FlowchartJsonConverter.cs:45",
						"notFoundConnections": [],
						"canStartWorkflow": false,
						"runAsynchronously": false
					},
					"activities": [
						{
							"result": null,
							"id": "71b62412efbfc6ca",
							"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d:a2307a0b4b92bb79:71b62412efbfc6ca",
							"name": "ErrorHandlerActivity1",
							"type": "Wave.Service.HostedServiceHandlers.ExternalEvents.WaveDrafted.ElsaWorkflow.ErrorHandlerActivity",
							"version": 1,
							"customProperties": {
								"canStartWorkflow": false,
								"runAsynchronously": false
							},
							"metadata": {
								"designer": {
									"position": {
										"x": -380.5,
										"y": -129
									},
									"size": {
										"width": 184.625,
										"height": 50
									}
								}
							}
						}
					],
					"connections": []
				},
				"result": null,
				"id": "7d8083b550abe76d",
				"nodeId": "Workflow1:626874ef88bcf26c:7d8083b550abe76d",
				"name": "If1",
				"type": "Elsa.If",
				"version": 1,
				"customProperties": {
					"canStartWorkflow": false,
					"runAsynchronously": false
				},
				"metadata": {
					"designer": {
						"position": {
							"x": 540,
							"y": 320
						},
						"size": {
							"width": 376.453125,
							"height": 120
						}
					}
				}
			}
		],
		"connections": [
			{
				"source": {
					"activity": "5212ab931be19aee",
					"port": "Done"
				},
				"target": {
					"activity": "7d8083b550abe76d",
					"port": "In"
				}
			}
		]
	}
}

Update- Added customActivity

 public class ActivityWithOutput : CodeActivity<string>
 {
     public ActivityWithOutput( )
     { 
     }

     protected override void Execute(ActivityExecutionContext context)
     { 
         var result = "My test output";
         context.SetResult(result);
     }
 }

**Comment by glime-ai** [comment]
Based on the information provided, it seems like you are encountering an error related to accessing workflow input within an "Elsa.FlowDecision" activity using a C# expression. The error message indicates that the memory block reference for the expression does not exist. This could be due to several reasons, such as the workflow not being properly initialized with the expected inputs or outputs, or there might be an issue with the way the expression is written or evaluated.

To troubleshoot this issue, you can check the following:

1. Ensure that the workflow is correctly receiving and initializing the inputs it expects. If the workflow expects certain inputs, they need to be provided when the workflow is triggered.

2. Verify that the expression used in the "condition" field of the "If1" activity is correct. The expression `(string)Output.LastResult == "test"` is trying to cast the last result of an activity to a string and compare it with "test". Make sure that the `Output.LastResult` is actually producing a result and that it is of the correct type.

3. If you are using custom activities (as it seems from the activity types like `Wave.Service.HostedServiceHandlers.ExternalEvents.WaveDrafted.ElsaWorkflow.ActivityWithOutput`), ensure that they are correctly implemented and registered within the Elsa workflow environment.

4. Review the Elsa documentation and examples to ensure that your workflow definition follows the correct format and conventions for defining activities, expressions, and connections.

5. If the issue persists, consider reaching out to the Elsa community or support channels with detailed information about your workflow and the specific error message you are encountering.

Unfortunately, without access to the actual implementation of your custom activities and the full context of your workflow environment, it's challenging to provide a more specific solution. However, the steps above should help you in identifying and resolving the issue.

If you need further assistance, please provide additional details about the custom activities and the context in which the workflow is executed.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=dRbvTb7IhCSYiI&utm_medium=github).
